http.go

```diff
@@ -106,6 +117,45 @@ func InstrumentHandlerFunc(label string, f func(http.ResponseWriter, *http.Reque
 	return InstrumentHandler(label, http.HandlerFunc(f))
 }
 
+// traceSampler is a time-based request sampler for tracing.
+//
+// It allows at most one request to be traced during each time interval.
+type traceSampler struct {
+	intervalNs float64
+
+	mu               sync.Mutex
+	rng              *rand.Rand
+	nextSampleTimeNs atomic.Int64
+}
+
+// newTraceSampler returns a new traceSampler that allows at most one request
+// to be traced during each time interval.
+func newTraceSampler(interval time.Duration) *traceSampler {
+	return &traceSampler{
+		intervalNs: float64(interval / time.Nanosecond),
+		rng:        rand.New(rand.NewSource(time.Now().UnixNano())),
+	}
+}
+
+// shouldTrace returns true iff a request should be traced.
+func (s *traceSampler) shouldTrace() bool {
+	now := time.Now().UnixNano()
+	if now < s.nextSampleTimeNs.Load() {
+		return false
+	}
+
+	// Check again while holding the lock.
+	s.mu.Lock()
+	defer s.mu.Unlock()
+	if now < s.nextSampleTimeNs.Load() {
+		return false
+	}
+
+	// Pick a new value for nextSampleTime and return.
+	s.nextSampleTimeNs.Store(now + int64(2*s.rng.Float64()*s.intervalNs))
```

Comment:
Add comment:
   // Pick a random value in [0,s*s.intervalNs] (averages out to s.intervalNs).